src/gui/qgsgeometrywidget.cpp

```diff
@@ -0,0 +1,254 @@
+{
+  mAcceptedTypes = types;
+}
+
+QList<QgsWkbTypes::Type> QgsGeometryWidget::acceptedWkbTypes() const
+{
+  return mAcceptedTypes;
+}
+
+bool QgsGeometryWidget::isReadOnly() const
+{
+  return mReadOnly;
+}
+
+void QgsGeometryWidget::setReadOnly( bool readOnly )
+{
+  mReadOnly = readOnly;
+}
+
+void QgsGeometryWidget::clearGeometry()
+{
+  if ( mGeometry.isNull() )
+    return;
+
+  setGeometryValue( QgsReferencedGeometry() );
+}
+
+void QgsGeometryWidget::copyAsWkt()
+{
+  if ( mGeometry.isNull() )
+    return;
+
+  QApplication::clipboard()->setText( mGeometry.asWkt() );
+}
+
+void QgsGeometryWidget::copyAsGeoJson()
+{
+  if ( mGeometry.isNull() )
+    return;
+
+  QApplication::clipboard()->setText( mGeometry.asJson() );
+}
+
+void QgsGeometryWidget::pasteTriggered()
+{
+  if ( !mPastedGeom.isNull() )
+  {
+    setGeometryValue( mPastedGeom );
+    mPastedGeom = QgsReferencedGeometry();
+  }
+}
+
+void QgsGeometryWidget::fetchGeomFromClipboard()
+{
+  mPastedGeom = QgsReferencedGeometry();
+  if ( mReadOnly )
+    return;
+
+  const QString text = QApplication::clipboard()->text();
+  if ( text.isEmpty() )
+    return;
+
+  //try reading as a single wkt string
+  mPastedGeom = QgsReferencedGeometry( QgsGeometry::fromWkt( text ), QgsCoordinateReferenceSystem() );
+  if ( !mPastedGeom.isNull() && typeIsAcceptable( mPastedGeom.wkbType() ) )
+  {
+    return;
+  }
+  mPastedGeom = QgsReferencedGeometry();
+
+  //try reading as a list
+  const QStringList lines = text.split( "\n", Qt::SkipEmptyParts );
+  if ( !lines.isEmpty() )
+  {
+    for ( const QString &line : lines )
+    {
+      const QgsGeometry geometry = QgsGeometry::fromWkt( line );
+      if ( !geometry.isNull() && typeIsAcceptable( geometry.wkbType() ) )
+      {
+        mPastedGeom = QgsReferencedGeometry( geometry, QgsCoordinateReferenceSystem() );
+        return;
+      }
+    }
+  }
+
+  // try reading a GeoJSON
+  const QgsFeatureList features = QgsJsonUtils::stringToFeatureList( text );
+  if ( !features.isEmpty() && features.at( 0 ).hasGeometry() )
+  {
+    mPastedGeom = QgsReferencedGeometry( features.at( 0 ).geometry(), QgsCoordinateReferenceSystem() );
+    return;
+  }
+}
+
+bool QgsGeometryWidget::typeIsAcceptable( QgsWkbTypes::Type type ) const
+{
+  if ( mAcceptedTypes.isEmpty() )
+  {
+    return true;
+  }
+
+  return mAcceptedTypes.contains( type );
+}
+
+void QgsGeometryWidget::prepareMenu()
+{
+  fetchGeomFromClipboard();
+
+  mCopyGeoJsonAction->setEnabled( !mGeometry.isNull() );
+  mCopyWktAction->setEnabled( !mGeometry.isNull() );
+  mClearAction->setEnabled( !mGeometry.isNull() && !mReadOnly );
+  mPasteAction->setEnabled( !mPastedGeom.isNull() && !mReadOnly );
+}
+
+void QgsGeometryWidget::updateLineEdit()
+{
+  if ( mGeometry.isNull() )
+  {
+    mLineEdit->setText( QgsApplication::nullRepresentation() );
+  }
+  else
+  {
+    QString wkt = mGeometry.asWkt();
+    if ( wkt.length() >= 1050 )
+    {
+      wkt = wkt.left( 999 ) + QChar( 0x2026 );
```

Comment:
May I suggest to use a constant variable for that max lenght ? Does it make sense to have such a large limit, btw ?

Reply:
Ok, I've added this to QgsField so that it can be used by the wkt trimming logic in there too